be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+					      struct m0_be_seg	     *seg,
+					      struct m0_be_tx_credit *accum)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_CREDIT_PTR(rec, seg, accum);
+	M0_BE_ALLOC_CREDIT_BUF(payload, seg, accum);
+	dtm0_be_alloc_tx_desc_credit(txd, seg, accum);
+	lrec_be_list_credit(M0_BLO_TLINK_CREATE, 1, accum);
+}
+
+M0_INTERNAL void dtm0_be_set_log_rec_credit(struct m0_dtm0_tx_desc *txd,
+					     struct m0_buf	    *payload,
+					     struct m0_be_seg	    *seg,
+					     struct m0_be_tx_credit *accum)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_CREDIT_BUF(payload, seg, accum);
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*payload));
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*txd));
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa *)));
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa)));
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*rec));
+}
+
+M0_INTERNAL void dtm0_be_del_log_rec_credit(struct m0_be_seg       *seg,
+					    struct m0_dtm0_log_rec *rec,
+					    struct m0_be_tx_credit *accum)
+{
+	lrec_be_list_credit(M0_BLO_TLINK_DESTROY, 1, accum);
+	M0_BE_FREE_CREDIT_ARR(rec->dlr_txd.dtd_ps.dtp_pa,
+			      rec->dlr_txd.dtd_ps.dtp_nr, seg, accum);
+	M0_BE_FREE_CREDIT_PTR(rec->dlr_payload.b_addr, seg, accum);
+	M0_BE_FREE_CREDIT_PTR(rec, seg, accum);
+}
+
+M0_INTERNAL void dtm0_be_log_destroy_credit(struct m0_be_seg       *seg,
+					    struct m0_be_tx_credit *accum)
+{
+	lrec_be_list_credit(M0_BLO_DESTROY, 1, accum);
+}
+
+M0_INTERNAL void m0_be_dtm0_log_credit(enum m0_be_dtm0_log_credit_op op,
+				       struct m0_dtm0_tx_desc	    *txd,
+				       struct m0_buf		    *payload,
+                                       struct m0_be_seg             *seg,
+				       struct m0_dtm0_log_rec	    *rec,
+                                       struct m0_be_tx_credit       *accum)
+{
+	/*TODO:Complete implementation during persistent list implementation */
+	struct m0_be_dtm0_log *log;
+
+	switch (op) {
+	case M0_DTML_CREATE:
+		M0_BE_ALLOC_CREDIT_PTR(log, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->dl_seg, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->dl_list, seg, accum);
+		lrec_be_list_credit(M0_BLO_CREATE, 1, accum);
+		m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_PTR(log));
+		break;
+	case M0_DTML_EXECUTED:
+	case M0_DTML_PERSISTENT:
+		dtm0_be_alloc_log_rec_credit(txd, payload, seg, accum);
+		dtm0_be_set_log_rec_credit(txd, payload, seg, accum);
+		m0_be_list_credit(M0_BLO_ADD, 1, accum);
+		break;
+	case M0_DTML_PRUNE:
+		dtm0_be_del_log_rec_credit(seg, rec, accum);
+		m0_be_list_credit(M0_BLO_DEL, 1, accum);
+		break;
+	case M0_DTML_REDO:
+	default:
+		M0_IMPOSSIBLE("");
+	}
+}
+
+M0_INTERNAL int m0_be_dtm0_log_create(struct m0_be_tx        *tx,
+                                      struct m0_be_seg       *seg,
+                                      struct m0_be_dtm0_log **out)
```

Comment:
What is the difference between this function and "m0_be_dtm0_log_init()"?

Please add comments for these functions.

Reply:
Sergey please write about semantics of init/create, where init initializes volatile fields of data structure, and create creates persistent.

Reply:
I have refactored the code from m0_be_dtm0_log_init() and moved the portion where we alloc memory for a volatile log into a separate function m0_be_dtm0_log_alloc(). So, now things are relatively  simple. m0_be_dtm0_log_alloc() should be called for allocating and in-memory/volatile memory log. For creating a persistent/on-disk log we need to call m0_be_dtm0_log_create.